### PR TITLE
Support modifying the plot style by introducing a generic framework for overriding components

### PR DESCRIPTION
**Pull request by jleibs**
### What
- Part of https://github.com/rerun-io/rerun/issues/4817

This tackles the problem of allowing SeriesStyle components to be overridden by building out a significant chunk of a general component override framework.

The new UI is limited to TimeSeriesScalar view at the moment because it still requires special Visualizer-side handling of the component data.

### High level idea
When selecting an entity within a space-view, we now have the ability to override any component that is both queried by the visualizer, and for which we have a registered component-editor.

This override value is initialized based on looking for, in order:
 - The current value of the component for the entity
 - A value provided by the Visualizer (useful for things like matching the current auto-color)
 - A default value provided by a function registered along with the editor.

### Notable changes:
 - Adds a new mechanism to the Component UI Registry that allows us to register editors and default-value providers for any component.
 - Adds a mechanism to all visualizers to find out the full set of components they query.
 - During the `update_overrides` process, we now checks the appropriate location in the blueprint override tree and insert the path to any overridden components into the DataResult.
 - In the TimeSeriesSpaceView we now respect any overrides that come from the DataResult.
 - Introduce a new Override Components UI in the seleciton panel, which allows us to add an override using a default-provided, or visualizer-provided initial value.
 - Once the override has been added, the override UI dispatches to the registered editor for each overridden component.

### Screenshot
image

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4914/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4914/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4914/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4914)
- [Docs preview](https://rerun.io/preview/e5f379759f9e985910550ab0258abdec4d4ca1a1/docs) 
- [Examples preview](https://rerun.io/preview/e5f379759f9e985910550ab0258abdec4d4ca1a1/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_viewer/src/ui/override_ui.rs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        ui.style_mut().wrap = Some(false);
+
+        let view_systems = ctx
+            .space_view_class_registry
+            .new_visualizer_collection(*space_view.class_identifier());
+
+        // We have to have at least 1 visualizer system or we can't create an override
+        let Some(system_for_initial_value) = view_systems.systems.values().next() else {
+            return;
+        };
```
Comment (Wumpf) [inline crates/re_viewer/src/ui/override_ui.rs:171]: so if we have several systems that could provide an initial valuewe just ignore those?

File: crates/re_viewer/src/ui/override_ui.rs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+        let mut components_per_visualizer = data_result
+            .visualizers
+            .iter()
+            .filter_map(|vis| view_systems.get_by_identifier(*vis).ok())
+            .map(|vis| vis.visualizer_query_info().queried);
+
+        // Accmulate all the components if there are multiple visualizers
```
Comment (Wumpf) [inline crates/re_viewer/src/ui/override_ui.rs:179]:
```suggestion
        // Accumulate all the components if there are multiple visualizers
```

File: crates/re_viewer/src/ui/selection_panel.rs [hunk 3]
```diff
@@ -190,6 +193,28 @@ impl SelectionPanel {
                     });
                 }
 
+                // Special section for space-view-entities
+                if let Item::InstancePath(Some(space_view_id), instance_path) = item {
+                    if instance_path.instance_key == InstanceKey::SPLAT {
+                        if let Some(space_view) = viewport.blueprint.space_views.get(space_view_id)
+                        {
+                            // TODO(jleibs): Overrides still require special handling inside the space view.
```
Comment (Wumpf) [inline crates/re_viewer/src/ui/selection_panel.rs:201]: technically not inside the space view but inside the affected visualizers

File: crates/re_viewer/src/ui/override_ui.rs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        let mut components = components_per_visualizer.next().unwrap_or_default();
+        for mut rest in components_per_visualizer {
+            components.append(&mut rest);
+        }
+
+        // Empty space views of every available types
```
Comment (Wumpf) [inline crates/re_viewer/src/ui/override_ui.rs:185]: space views? I presume you mean components

File: crates/re_viewer/src/ui/override_ui.rs [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        for component in components.difference(active_components) {
+            // If we don't have an override_path we can't set up an initial override
+            // this shouldn't happen if the `DataResult` is valid.
+            let Some(override_path) = data_result.override_path() else {
+                continue;
```
Comment (Wumpf) [inline crates/re_viewer/src/ui/override_ui.rs:190]: so it would be an error that we should log?

File: crates/re_viewer/src/ui/override_ui.rs [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                    .and_then(|result| result.2[0].clone())
+                    .and_then(|cell| {
+                        if cell.num_instances() == 1 {
+                            Some(cell)
+                        } else {
+                            None
+                        }
+                    })
+                    .or_else(|| {
+                        system_for_initial_value.initial_override_value(
```
Comment (Wumpf) [inline crates/re_viewer/src/ui/override_ui.rs:220]: this doesn't check out for me: The system for the initial value may not have brought up the component that we're on right now.
I think what you iterate instead is pairs of component-visualizer where you somehow removed duplicated components, i.e. decided which visualizer would provide the initial override value :/

File: crates/re_viewer/src/ui/override_ui.rs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                            &data_result.entity_path,
+                            component,
+                        )
+                    })
+                else {
+                    re_log::warn!("Could not identify an initial value for: {}", component);
+                    return;
+                };
+
+                initial_data.compute_size_bytes();
```
Comment (Wumpf) [inline crates/re_viewer/src/ui/override_ui.rs:242]: why is this needed?

File: crates/re_viewer/src/ui/selection_panel.rs [hunk 8]
```diff
@@ -46,7 +49,7 @@ pub(crate) struct SelectionPanel {
 /// timeline, then use the blueprint. Otherwise use the recording.
 // TODO(jleibs): Ideally this wouldn't be necessary and we could make the assessment
 // directly from the entity_path.
-fn guess_query_and_store_for_selected_entity<'a>(
+pub(crate) fn guess_query_and_store_for_selected_entity<'a>(
```
Comment (Wumpf) [inline crates/re_viewer/src/ui/selection_panel.rs:52]: I suppose this partially comes back to "can override paths exist in the data store" if the answer is yes this becomes both necessary and fragile. If no, we can check if an path is an override and pick the right store making this not a "guess" but a "know"

File: crates/re_viewer/src/ui/override_ui.rs [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+                                &query,
+                                store,
+                                entity_path,
+                                &overrides.override_path,
+                                &component_data,
+                                instance_key,
+                            );
+                        } else {
+                            // TODO(jleibs): This shouldn't happen. Warn instead?
+                            ui.weak("(empty)");
```
Comment (Wumpf) [inline crates/re_viewer/src/ui/override_ui.rs:144]: we have `re_ui.warning_text` for cases like this - this could then print out what's happening, even with a longer explanation in the tooltip

**Review by Wumpf — APPROVED** [review]
absolutely marvelous and not as complex as I feared it would be.

Comments are mostly starts of discussions, some of which we should have over Zoom or Slack instead, some of which need to be postponed.
The only thing that I believe is actually wrong is  how initial overrides are determined (see comments). But since this won't affect the time series space view, I'll approve anyways (:

File: crates/re_data_ui/src/editors.rs [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    components::{Color, Radius, ScalarScattering, Text},
+    Component, Loggable,
+};
+use re_viewer_context::{UiVerbosity, ViewerContext};
+
+#[allow(clippy::too_many_arguments)]
+fn edit_color_ui(
+    ctx: &ViewerContext<'_>,
+    ui: &mut egui::Ui,
+    _verbosity: UiVerbosity,
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_data_ui/src/editors.rs:17]: I think it will eventually. I suspect there will be both a Small/Inline editor for some things, when it's appropriate, but I could imagine a way of expanding that to a more "full" multiline editor that we won't always want to show in the selection panel by default.

File: crates/re_space_view_time_series/src/visualizer_system.rs [hunk 11]
```diff
@@ -190,6 +206,19 @@ impl TimeSeriesSystem {
...
+                    egui::Color32::from_rgba_unmultiplied(arr[0], arr[1], arr[2], arr[3])
+                });
+
+                let override_label =
+                    lookup_override::<Text>(data_result, ctx).map(|t| t.to_string());
+
+                let override_scattered =
+                    lookup_override::<ScalarScattering>(data_result, ctx).map(|s| s.0);
+
+                let override_radius = lookup_override::<Radius>(data_result, ctx).map(|r| r.0);
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_space_view_time_series/src/visualizer_system.rs:220]: Yeah, it's not awful, esp since we restrict things to splats at the moment, but I'd still love to have it automagically happen inside one of the archetype-view-style generic queries.

File: crates/re_viewer_context/src/space_view/view_query.rs [hunk 12]
```diff
@@ -23,6 +24,12 @@ pub struct PropertyOverrides {
     /// The individual property set in this `DataResult`, if any.
     pub individual_properties: Option<EntityProperties>,
 
+    /// An alternative store and entity path to use for the specified component.
+    // NOTE: StoreKind is easier to work with than a `StoreId`` or full `DataStore` but
+    // might still be ambiguous when we have multiple stores active at a time.
+    // TODO(jleibs): Consider something like `tinymap` for this.
+    pub component_overrides: HashMap<ComponentName, (StoreKind, EntityPath)>,
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer_context/src/space_view/view_query.rs:31]: Maybe overrides is the wrong word to use here. Could instead call it something like "resolved_component_location"

Imagine that we configured an entity to inherit a list of components from its parent (or another linked entity) this let's us represent that reference.

File: crates/re_viewer_context/src/space_view/visualizer_system.rs [hunk 13]
```diff
@@ -66,6 +93,19 @@ pub trait VisualizerSystem: Send + Sync + 'static {
     }
 
     fn as_any(&self) -> &dyn std::any::Any;
+
+    /// Returns an initial value to use when creating an override for a component for this
+    /// visualizer. This is used as a fallback if the component doesn't already have data.
+    fn initial_override_value(
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer_context/src/space_view/visualizer_system.rs:99]: Yeah, would love to avoid the overhead here. Type-erasure is such a pain. The generic UI also needs to be able to dispatch to it based on ComponentName. Would love to brainstorm some ways to make this path cleaner.

File: crates/re_viewer_context/src/component_ui_registry.rs [hunk 14]
```diff
@@ -46,26 +47,68 @@ type ComponentUiCallback = Box<
         + Sync,
 >;
 
+type ComponentEditCallback = Box<
+    dyn Fn(
+            &ViewerContext<'_>,
+            &mut egui::Ui,
+            UiVerbosity,
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer_context/src/component_ui_registry.rs:54]: Happy to make the change in a separate PR -- let's discuss this one as we're building out more editors.

File: crates/re_data_ui/src/editors.rs [hunk 15]
```diff
@@ -0,0 +1,221 @@
+// TODO(jleibs): Turn this into a trait
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_data_ui/src/editors.rs:1]: The same way we have a DataUi trait, an EditUi trait seems like it would be a bit cleaner than throwing around bare callbacks.

File: crates/re_viewer/src/ui/override_ui.rs (hunk 1 above)
Comment (jleibs) [reply crates/re_viewer/src/ui/override_ui.rs:171]: For now, yes. We could do something more complex in the future but holding off on this until we understand the context.

File: crates/re_viewer/src/ui/override_ui.rs (hunk 4 above)
Comment (jleibs) [reply crates/re_viewer/src/ui/override_ui.rs:185]: No idea what this comment was referring to -- maybe a bad copy-paste from somewhere :man_shrugging:

File: crates/re_viewer/src/ui/override_ui.rs (hunk 7 above)
Comment (jleibs) [reply crates/re_viewer/src/ui/override_ui.rs:242]: :shrug: it's always needed when we write data to the store or else we get errors about using data that hasn't been sized yet.

File: crates/re_viewer/src/ui/selection_panel.rs (hunk 8 above)
Comment (jleibs) [reply crates/re_viewer/src/ui/selection_panel.rs:52]: This is actually sort of the opposite -- not the location of the override, but the location of the overridden data.  However, there's no meaningful way (yet) to override data that lives in the blueprint store.

This is actually an unnecessary copy-paste from some other selection panel logic. Removed the guess part and added a comment.

File: crates/re_viewer/src/ui/override_ui.rs (hunk 9 above)
Comment (jleibs) [reply crates/re_viewer/src/ui/override_ui.rs:144]: I'm not totally confident it should be a warn though either. Changing the comment.

File: crates/re_viewer/src/ui/override_ui.rs (hunk 6 above)
Comment (jleibs) [reply crates/re_viewer/src/ui/override_ui.rs:220]: Aha! Yes, good catch. Changed to build a map of Component -> Visualizer and use the right now.
